### ISSUE TITLE
Cannot use import statement outside a module error after NX 16.10.0 upgrade

### ISSUE THREAD

**Issue by wcwcaseman** [nrwl/nx#28016]
### Current Behavior

During serve of my website my build fails. In webpack.config.ts I am loading a ts file using require. I have an import at the top of that ts file. I get the following error Cannot use import statement outside a module. I did not get this error before upgrading to NX 16.10.0. I was on 16.9.1.

### Expected Behavior

Should be able to load typescript file in webpack.config.ts without issue and the application should start.

### Steps to Reproduce

1. Create nx react mono repo version 16.10.0 or above
2. Upgrade @swc-node/register to latest or remove it, as otherwise another defect will occur. 
3. Add a ts file with an import statement
4. Load that ts file using require in webpack.config.ts
5. Run the application

### Nx Report

```shell
Node   : 18.18.2
   OS     : darwin-x64
   pnpm   : 8.10.5
   
   nx                 : 16.10.0
   @nx/js             : 16.10.0
   @nx/jest           : 16.10.0
   @nx/linter         : 16.10.0
   @nx/workspace      : 16.10.0
   @nx/cypress        : 16.10.0
   @nx/devkit         : 16.10.0
   @nx/eslint-plugin  : 16.10.0
   @nx/plugin         : 16.10.0
   @nx/react          : 16.10.0
   @nx/storybook      : 16.10.0
   @nrwl/tao          : 16.10.0
   @nx/web            : 16.10.0
   @nx/webpack        : 16.10.0
   typescript         : 5.1.6
   ---------------------------------------
   Community plugins:
   ts-check : 1.0.0
   ---------------------------------------
   Local workspace plugins:
         @ecp/workspace-plugin
```

### Failure Logs

```shell
Cannot use import statement outside a module


/Users/myfile.ts:1
import { item } from 'items';

SyntaxError: Cannot use import statement outside a module
```

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

**Comment by wcwcaseman** [comment]
@jaysoo I upgraded to 19.8.1, With swc libraries I get a different error context is undefined. So I removed swc again. It looks as maybe this issue is partially fixed. Now I get the error 'Cannot find module '@example/payment' .  I have these paths in tsconfig.base.json, not sure why they aren't working anymore.

**Comment by wcwcaseman** [comment]
> reply to keep it active.

This is still an issue with the latest version of NX. Last time I tested with the latest version It wouldn't let me not use SWC so my code failed on path. I couldn't even get to this error because path errors.

**Comment by wcwcaseman** [comment]
> [@wcwcaseman](https://github.com/wcwcaseman) Can you retry with latest Nx (e.g. v19) and see if the issue persists?
> 
> Nx v16 will be out of LTS next month. https://nx.dev/reference/releases

Not sure if I was supposed to reply this way, but yes this is still and issue.

**Comment by wcwcaseman** [comment]
Tested with NX version 20 and node version 20.17 with SWC or without SWC, path.resolve returns TypeError: Cannot read properties of undefined (reading 'resolve').

**Comment by wcwcaseman** [comment]
@jaysoo I know how to fix this can you help me? I'm not sure how to make a MR to version 17.0.X.

**Comment by wcwcaseman** [comment]
@jaysoo I was able to open a merge request, please take a look.

**Comment by Coly010** [comment]
@wcwcaseman it seems like the correct fix here is for us to make sure the tsconfig.base.json is loaded so that TS Path aliases are handled. 

This should already be the case, but I’ll double check it.

However, it is unlikely that we’ll release a patch for this for v17.

**Comment by g3tr1ght** [comment]
@Coly010 getting traction on this issue is great, thank you for that!

**TLDR**
- **v17 and upwards**: path aliases mapper started via registerTsConfigPaths is turned off (prematurely) right after loading custom webpack config,
- **v17-v19.3.x**: transpiler started via registerTranspiler is turned off (prematurely) right after loading custom webpack config

The problem is applicable to both ts-node and swc-node.

Pretty much the whole thing is happening in https://github.com/nrwl/nx/blob/20.0.x/packages/nx/src/plugins/js/utils/register.ts and as the module location hasn't changed between v17 and latest, a fix potentially can be backported?

**Here's some analysis.**

In essence, the problem is not with tsconfig.base.json path aliases because NX executors are using project's tsconfig, which extends tsconfig.base.json, so paths are acknowledged by ts-node and swc-node. We understand that bundler itself might be using ts-loader, swc-loader, babel-loader depending on one's setup, and the problem is not there either.

The problem is two-fold and is specific to the NX's transpiler abstracted implementation that's running on the auxiliary scripts (like custom webpack config and all imports down to every downstream module, like when webpack config imports xyz.ts and xyz.ts imports abc.ts). NX implementation turns on transpiler before loading custom webpack config and turns it off right after, so it's able to transpile static imports only. NX then eventually runs webpack and if a custom config had a dynamic import (which happens only when webpack is actually running), it ends up trying to import non-transpiled module and crashing because this line turned it off: https://github.com/nrwl/nx/blob/20.0.x/packages/webpack/src/utils/webpack/resolve-user-defined-webpack-config.ts#L26 (same implementation in v17, only the file had a different name)
By transpiler I mean both transpiler itself and path aliases mapper (which patches node module resolution algo).

First part of the problem was fixed by making ts-node transpiler cleanup function a no-op in 19.4.0 release: https://github.com/nrwl/nx/commit/b2feceae4f56442352505f5f4cb5678050ace329#diff-ba45dda1a4c0ffc9b5da8dc11d88a7e019e113d06639286f61c09601820eca19R152 and all succeeding releases look good from this perspective.

However, there is still another problem related to path aliases stemming from about the same location in NX source code. Along with transpiler itself, NX adds path aliases mapper https://github.com/nrwl/nx/blob/20.0.x/packages/nx/src/plugins/js/utils/register.ts#L84-L87 https://github.com/nrwl/nx/blob/20.0.x/packages/nx/src/plugins/js/utils/register.ts#L380, and as part of the cleanup function https://github.com/nrwl/nx/blob/20.0.x/packages/webpack/src/utils/webpack/resolve-user-defined-webpack-config.ts#L26 https://github.com/nrwl/nx/blob/20.0.x/packages/nx/src/plugins/js/utils/register.ts#L110-L112, **disables the mapper** (tsconfig-paths reverts to the original module resolution algo https://github.com/dividab/tsconfig-paths/blob/master/src/register.ts#L121)

Potential fix is to keep the mapper active (no-op?), which would close this issue along with https://github.com/nrwl/nx/issues/26329. Ideally this fix along with https://github.com/nrwl/nx/commit/b2feceae4f56442352505f5f4cb5678050ace329#diff-ba45dda1a4c0ffc9b5da8dc11d88a7e019e113d06639286f61c09601820eca19R152 need to be backported down to v17.

**Comment by Coly010** [comment]
@g3tr1ght thank you for the detailed analysis. I can understand better here now why issues arise when relying on ts-path mappings. Dynamic imports do present an issue if they’re not eagerly evaluated when we create the config that is passed to Webpack api. 

as mentioned, however, it’s very unlikely this will get back ported. Other than this, what issues prevents migration to latest Nx?

**Comment by g3tr1ght** [comment]
@Coly010 I think it's reasonable to expect that imports can be fully dynamic (ones that can't be statically analyzable) when it comes to auxiliary scripts and node services so a fix would be very appreciated.

Fair point regarding backport, we can patch offending module during installation as a workaround while gradually migrating.
Regarding migrating to the latest, our repo is of decent size and has customizations on top of NX, so we can't make huge leaps and slowly incrementing NX version helps us understanding any important diffs, regressions (when/where/why they happened). A huge leap will only amplify regressions and hinder debugging in my opinion. Keeping all dependencies up to date is impossible due to other priorities taking over. However, as you suggested, we tried using latest version and it seem to successfully run our apps when removing cleanup function call, though I'm not sure if there's other issues to uncover but we'll get there eventually.

Thank you!

**Comment by g3tr1ght** [comment]
As a side note, I personally prefer to maintain a separate explicit tsconfig for tooling because our apps/libraries usually have different expectations for transpilation process compared with our tooling, just like it's done in NX with tsconfig.spec.json. So we could have had tsconfig.tools.json in maybe root of the repo? Currently, NX uses app/library tsconfig and alters some config entries during execution, which you could only know about if you review NX source code, also `--verbose` doesn't add much output because most of the important events aren't logged. Just food for thought.

**Comment by g3tr1ght** [comment]
not stale

**Comment by wcwcaseman** [comment]
Tested with latest 20.0.12, this is still an issue.

**Comment by g3tr1ght** [comment]
@Coly010 is it possible to add to the roadmap the fix for the issue? Or else are there any reasons that inability to use dynamic require/import in webpack related scripts is expected?
